### PR TITLE
Add support for the subsystem feature

### PR DESCRIPTION
**Pull request by rgrinberg**
Extracted from the inline tests PR.

### CODE REVIEW

**Comment by diml** [comment]
Yh, I suppose there are two kinds of versionning:
- one for the file itself
- one for each sub-system

These files are not meant to be written by the user, so we can use something simple:

```
(<file-spec-version>
  (<sub-system> <sub-system-version> <sub-system-data>))
```

For instance:

```
(1.0
 (inline_tests.backend 1.0 ((gen (...)) (flags (...)) (runner_libs (...)))))
```

**Comment by diml** [comment]
I removed the last commit as we are no longer generating the variable in META file

**Comment by rgrinberg** [comment]
I see that the new backends can now have dependencies on other backends, could you motivate which use cases you have in mind for this?

**Comment by diml** [comment]
> I moved the last commit to master just to keep things tidy. I get this error when running the tests:

Yh, I fixed it but I'm not sure why this is not affecting master. BTW, we should add flag in `Exe.build_and_link` to allow disabling symlinking the cm* files. There is no point symlinking them for implicit executables such as utop or  inline tests runner.

**Comment by diml** [comment]
> I see that the new backends can now have dependencies on other backends, could you motivate which use cases you have in mind for this?

So the interface say dependencies, but in practice the sub-system might interpret it differently. For inline tests, this allow one backend to extend another one. For instance ppx_expect extends ppx_inline_test.

For ppx_driver, we can use it to state that a driver replaces another one. For instance ppx_driver will say that it replaces ocaml-migrate-parsetree.

From the point of view of the `Sub_system` module it's all the same, all it needs to know is the relations between the various backends so that it can decide when things are compatible or not. For instance, using both qtest and ppx_inline_test is not allowed as they don't compose, but using both ppx_inline_test and ppx_expect is.

**Comment by rgrinberg** [comment]
> BTW, we should add flag in Exe.build_and_link to allow disabling symlinking the cm* files. There is no point symlinking them for implicit executables such as utop or inline tests runner.

Is this just for performance? If it is, it doesn't seem all that worthwhile to complicate things unless building runners can ever be a bottleneck.

**Comment by diml** [comment]
No, it's because otherwise they get picked up by things like `(glob_files src/*.cmi)`

**Comment by diml** [comment]
Considering this and https://github.com/ocaml/dune/pull/548#issuecomment-368301753, I'm thinking that we shouldn't have moved utop in the main library directory. I think we should revert the change and also generate the inline test runners in separate sub-directories to avoid similar problems. One thing we can do to simplify things a bit, is set the object directory to just the `.utop` directory, since the cm* files will already be isolated.

File: src/installed_dune_file.ml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  |> Sub_system_name.Map.of_list
+  |> (function
+    | Ok x -> x
+    | Error (name, _, (loc, _, _)) ->
+      Loc.fail loc "%S present twice" (Sub_system_name.to_string name))
+  |> Sub_system_name.Map.mapi ~f:(fun name (_, version, data) ->
+    let (module M) = Jbuild.Sub_system_info.get name in
+    let vloc, ver = version in
+    if M.version <> ver then
+      Loc.fail vloc "Unsupported version, only version %s is supported"
```
Comment (rgrinberg) [inline src/installed_dune_file.ml]: I'd include `ver` here as well. Even though it's accessible via `vloc`, it's still an annoying roundtrip.

Comment (diml) [reply src/installed_dune_file.ml]: This code has moved to a new `Syntax.Versioned_parser` module and I included the version number in the message

**Comment by diml** [comment]
The installed dune files are now properly versioned as well as each sub-system

File: src/gen_rules.ml [hunk 2]
```diff
@@ -805,6 +844,10 @@ module Gen(P : Params) = struct
      | Aliases                                                         |
      +-----------------------------------------------------------------+ *)
 
+  let interpret_locks ~dir ~scope locks =
+    List.map locks ~f:(fun s ->
+      Path.relative dir (SC.expand_vars sctx ~dir ~scope s))
+
```
Comment (rgrinberg) [inline src/gen_rules.ml]: This function already exists in this file, is this a typo?

File: src/sub_system.mli [hunk 3]
```diff
@@ -0,0 +1,11 @@
+open! Import
+
+include module type of struct include Sub_system_intf end
+
+module Register_backend(M : Backend) : Registered_backend with type t = M.t
+
+module Register_with_backend(M : With_backend) : sig end
```
Comment (rgrinberg) [inline src/sub_system.mli]: Could you leave a comment for each of these functors?

File: src/gen_rules.ml (hunk 2 above)
Comment (diml) [reply src/gen_rules.ml]: oops, removed

File: src/sub_system.mli (hunk 3 above)
Comment (diml) [reply src/sub_system.mli]: I added some doc. I also renamed `With_backend` as it wasn't a really good name

**Pull request merged**